### ISSUE TITLE
v1.2.3: rxmongo - MongoReadJournal.eventsByPersistenceId doesn't guarantee consecutive event history

### ISSUE THREAD

**Issue by marcuslinke** [scullxbones/akka-persistence-mongo#114]
@scullxbones Under some circumstances the eventstream subscription doesn't deliver `Event`s in ascending sequence number order. This leads to missing events in the view as the `RemoveDuplicatedEvents` stage drops events with sequence numbers lower than the current maximum.

**Comment by scullxbones** [comment]
Ok will need to back up and see what alternatives there are for event stream.  Good find!

**Comment by marcuslinke** [comment]
I suspect the insert into the capped collection as root cause. See https://github.com/scullxbones/akka-persistence-mongo/blob/master/rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala#L67

As the insert into the `realtime` collection is async it would need to chain it with the normal insert future somehow I guess. WDYT?

**Comment by marcuslinke** [comment]
After hacking of `batchAppend` in `RxMongoJournaller` like the following everything seems fine. Also the `RemoveDuplicatedEvents` stage can be removed then:

```scala
  private[mongodb] override def batchAppend(writes: ISeq[AtomicWrite])(implicit ec: ExecutionContext): Future[ISeq[Try[Unit]]] = {
    val batch = writes.map(aw => Try(driver.serializeJournal(Atom[BSONDocument](aw, driver.useLegacySerialization))))
    if (batch.forall(_.isSuccess)) {
      val collected = batch.toStream.collect { case Success(doc) => doc }
      val result = Failover2(driver.connection, driver.failoverStrategy) { () =>
        val j = journal.bulkInsert(collected, ordered = true, writeConcern).map(_ => batch.map(_.map(_ => ())))
        if (driver.realtimeEnablePersistence) {
          val r = realtime.bulkInsert(collected, ordered = true, writeConcern).map(_ => scala.collection.immutable.Seq[Try[Unit]]())
          Future.sequence(scala.collection.immutable.Seq(j, r)).map { _.flatten }
        } else {
          j
        }
      }.future
      
      result
    } else {
      Future.sequence(batch.foldLeft(scala.collection.immutable.Seq[Future[Try[Unit]]]()) {
        case (seq, Success(document: BSONDocument)) =>
          val result = journal.insert(document, writeConcern).map(writeResultToUnit)
          if (driver.realtimeEnablePersistence) {
            seq ++ Seq(result, realtime.insert(document, writeConcern).map(writeResultToUnit))
          } else seq :+ result
        case (seq, f: Failure[_]) => seq :+ Future.successful(Failure[Unit](f.exception))
      })
    }
  }
```

**Comment by scullxbones** [comment]
Thanks @marcuslinke - I had `Future.andThen` in mind, just because I don't want any `realtime` errors to be treated as `journal` errors.  But the fact that this fixes your issues is definitely good news as far as a solution approach.  It saves me from trying to replace the `eventStream` side.

**Comment by marcuslinke** [comment]
@scullxbones In contrast using `Future.andThen` would insert into both collections sequentially one after another. I'm not sure if this is desirable for performance reasons. However, would you like to fix it yourself or should I file a PR with the solution above?

**Comment by scullxbones** [comment]
@marcuslinke i'm definitely open to a PR - as long as it hits those couple requirements:

- Fixes the issue (of course)
- Handles the error case correctly, reporting a `journal` write error back to akka-persistence, even if `realtime` write succeeds

**Comment by marcuslinke** [comment]
OK. I will modify the code above using `Future.andThen`, do some final tests and create a PR tomorrow. Hopefully looking forward for a bufix release then... ;-)

**Comment by scullxbones** [comment]
I'll turn it around ASAP 😄

**Comment by marcuslinke** [comment]
@scullxbones Sadly it doesn't work with `Future.andThen`. Seems the inner future is not chained to the outer one then. Any ideas?

**Comment by marcuslinke** [comment]
OK. Sometimes the solution is so simple. Just using `flatMap` solves it for me: #115

**Comment by scullxbones** [comment]
Fixed - available in `1.2.4`.  Thanks for the contribution!

**Comment by marcuslinke** [comment]
Nice! Thanks for releasing. This (probably) saved my day. :smile:
